File: src/dashboard/utils/formatters.py
from __future__ import annotations

from typing import Optional

import pandas as pd


def format_qty(value: float, decimals: int = 6) -> str:
    """Format quantity with trimmed trailing zeros."""
    if pd.isna(value):
        return "N/A"
    text = f"{float(value):,.{decimals}f}"
    if "." in text:
        text = text.rstrip("0").rstrip(".")
    return text


def format_krw(value: float, signed: bool = False) -> str:
    """Format KRW amount with thousand separators."""
    if pd.isna(value):
        return "N/A"
    v = float(value)
    if signed:
        if v > 0:
            return f"+{v:,.0f}"
        if v < 0:
            return f"{v:,.0f}"
        return "0"
    return f"{v:,.0f}"


def format_krw_compact(value: float, signed: bool = False) -> str:
    """
    Compact KRW formatter.
    Example: 8,847,360 -> 884.7만 원
    """
    if pd.isna(value):
        return "N/A"

    v = float(value)
    abs_v = abs(v)
    if abs_v < 10000:
        return format_krw(v, signed=signed)

    compact = abs_v / 10000.0
    text = f"{compact:,.1f}만 원"

    if signed:
        if v > 0:
            return f"+{text}"
        if v < 0:
            return f"-{text}"
        return "0"

    if v < 0:
        return f"-{text}"
    return text


def format_pct(value: float, decimals: int = 2, signed: bool = True) -> str:
    """Format percentage with optional sign and zero without sign."""
    if pd.isna(value):
        return "N/A"
    v = float(value)
    if signed:
        if v > 0:
            return f"+{v:.{decimals}f}%"
        if v < 0:
            return f"{v:.{decimals}f}%"
        return f"{0:.{decimals}f}%"
    return f"{v:.{decimals}f}%"

File: frontend/next-dashboard/lib/formatters.js
/*
 * KRW 금액/퍼센트/수량 포맷터
 * Streamlit의 formatters.py와 동일한 형식을 재현한다.
 */

export function formatKrw(value, signed = false) {
  const n = Number(value) || 0;
  const abs = Math.abs(n);
  const formatted = abs.toLocaleString("ko-KR", {
    maximumFractionDigits: 0,
  });
  if (signed && n > 0) return `+${formatted}`;
  if (n < 0) return `-${formatted}`;
  return formatted;
}

export function formatKrwCompact(value, signed = false) {
  const n = Number(value) || 0;
  const abs = Math.abs(n);
  let text;
  if (abs >= 1_0000_0000) {
    text = `${(abs / 1_0000_0000).toFixed(1)}억`;
  } else if (abs >= 1_0000) {
    text = `${(abs / 1_0000).toFixed(1)}만`;
  } else {
    text = abs.toLocaleString("ko-KR");
  }
  if (signed && n > 0) return `+${text}`;
  if (n < 0) return `-${text}`;
  return text;
}

export function formatPct(value, decimals = 2, signed = true) {
  const n = Number(value);
  if (Number.isNaN(n)) return "N/A";
  const text = Math.abs(n).toFixed(decimals);
  if (signed && n > 0) return `+${text}%`;
  if (n < 0) return `-${text}%`;
  return `${text}%`;
}

export function formatQty(value) {
  const n = Number(value);
  if (Number.isNaN(n) || n === 0) return "0";
  return n.toFixed(8).replace(/0+$/, "").replace(/\.$/, "");
}
